### ISSUE TITLE
RPI ZERO 2W: don't works

### ISSUE THREAD

**Issue by qrp73** [seemoo-lab/nexmon#619]
New RPI ZERO 2W uses chip BCM43430/1, it using brcm/brcmfmac43436s-sdio firmware. 
It appears that there is no patch for this firmware...

I tried to rename patched brcmfmac43436-sdio to brcmfmac43436s-sdio, but it failed to load with error:
```
brcmfmac mmc1:0001:1: Direct firmware load for brcm/brcmfmac43436s-sdio.raspberrypi,model-zero-2-w.bin failed with error -2
```

**Comment by jlinktu** [comment]
We are not related to Raspberry Pi nor Kali Linux - but if you want to get this to work I suggest you to first get Kali Linux to correctly load the original firmware. Once this works, you can think about adding monitor mode.

**Comment by qrp73** [comment]
I just want to get working monitor mode, since it don't works on raspi-os, this is why I installed Kali Linux.

Original raspi-os-bookworm 6.6.31+rpt-rpi-v8  running on rpi2w uses this symlink ```/lib/firmware/brcm/brcmfmac43430-sdio.bin -> ../cypress/cyfmac43430-sdio.bin```

Here is log file from raspi-os-bookworm:
```
[   12.919275] brcmfmac: F1 signature read @0x18000000=0x1541a9a6
[   12.936804] brcmfmac: brcmf_fw_alloc_request: using brcm/brcmfmac43430-sdio for chip BCM43430/1
[   12.937728] usbcore: registered new interface driver brcmfmac
[   13.192057] brcmfmac: brcmf_c_process_txcap_blob: no txcap_blob available (err=-2)
[   13.192698] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM43430/1 wl0: Jun 14 2023 07:27:45 version 7.45.96.s1 (gf031a129) FWID 01-70bd2af7 es7
```

**Comment by jlinktu** [comment]
Well, if it doesn't matter what OS you are using, I suggest to go for Raspberry Pi OS then.

From the log output I can see that on Raspberry Pi OS, other than you suggested, [`brcmfmac43436s.bin`](https://github.com/RPi-Distro/firmware-nonfree/blob/bookworm/debian/config/brcm80211/brcm/brcmfmac43436s-sdio.bin) is loaded.

Regarding the confusion about what file is to be loaded, there are a couple of discussions there: https://github.com/RPi-Distro/firmware-nonfree/issues
TLDR: There are different versions of the Raspberry Pi Zero 2 W, with different Wi-Fi chips, which require different firmwares.
Apparently you have the version that requires [`brcmfmac43436s.bin`](https://github.com/RPi-Distro/firmware-nonfree/blob/bookworm/debian/config/brcm80211/brcm/brcmfmac43436s-sdio.bin). Kali simply seems to load the wrong firmware, thus, again suggesting to go with Raspberry Pi OS or port the related stuff to Kali.

However, all of the above has nothing to do with `nexmon` itself.
It is possible to patch [`brcmfmac43436s.bin`](https://github.com/RPi-Distro/firmware-nonfree/blob/bookworm/debian/config/brcm80211/brcm/brcmfmac43436s-sdio.bin) to support monitor mode, but I currently can't do this in my free-time. If you really need this, you can contact me by mail and we can try to find a solution.

**Comment by qrp73** [comment]
> Apparently you have the version that requires [`brcmfmac43436s.bin`](https://github.com/RPi-Distro/firmware-nonfree/blob/bookworm/debian/config/brcm80211/brcm/brcmfmac43436s-sdio.bin). Kali simply seems to load the wrong firmware, thus, again suggesting to go with Raspberry Pi OS or port the related stuff to Kali.

Why brcmfmac43436s-sdio??? 

When I boot into original Raspi OS with working wifi firmware it shows in the log that the chip is ```BCM43430/1``` and load firmware from brcmfmac43430-sdio which is symlink to ```/lib/firmware/cypress/cyfmac43430-sdio.bin``` and this firmware version is 7.45.96.s1 (gf031a129) FWID 01-70bd2af7 es7
 

> However, all of the above has nothing to do with `nexmon` itself. It is possible to patch [`brcmfmac43436s.bin`](https://github.com/RPi-Distro/firmware-nonfree/blob/bookworm/debian/config/brcm80211/brcm/brcmfmac43436s-sdio.bin) to support monitor mode, but I currently can't do this in my free-time. If you really need this, you can contact me by mail and we can try to find a solution.

Currently I'm trying to setup clean Raspi-OS Lite and make the patch. If I understand correctly there is needs to build ```patches/bcm43430a1/7_45_41_46/nexmon/``` and then try to replace original firmware with version 7.45.96.s1 at ```/lib/firmware/cypress/cyfmac43430-sdio.bin```. Is it correct?

**Comment by jlinktu** [comment]
The log says:
`brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM43430/1 wl0: Jun 14 2023 07:27:45 version 7.45.96.s1 (gf031a129) FWID 01-70bd2af7 es7`, which is the versioning string of [brcmfmac43436s-sdio.bin](https://github.com/RPi-Distro/firmware-nonfree/blob/bookworm/debian%2Fconfig%2Fbrcm80211%2Fbrcm%2Fbrcmfmac43436s-sdio.bin).

**Comment by qrp73** [comment]
just checked, yes - ```brcmfmac43436s-sdio.bin``` contains version string ```7.45.96.s1 (gf031a129)``` 

> It is possible to patch [`brcmfmac43436s.bin`](https://github.com/RPi-Distro/firmware-nonfree/blob/bookworm/debian/config/brcm80211/brcm/brcmfmac43436s-sdio.bin) to support monitor mode, but I currently can't do this in my free-time. If you really need this, you can contact me by mail and we can try to find a solution.

Thanks , I wrote you mail about it.

**Comment by alexzaporozhets** [comment]
@qrp73, @jlinktu any updates on this issue?

**Comment by qrp73** [comment]
@alexzaporozhets no, adding patch for 7.45.96.s1 is a paid service. 

But I found that old firmware BCM43430/1 version 7.45.41.46 (r666254 CY) which was supplied for RPI3 also works for RPI02W. And there is a patch in nexmon for this firmware version. You can found original this firmware version in this package: http://archive.raspberrypi.org/debian/pool/main/f/firmware-nonfree/firmware-brcm80211_0.43+rpi6_all.deb

**Comment by rudyrdx** [comment]
> @alexzaporozhets no, adding patch for 7.45.96.s1 is a paid service.
> 
> But I found old firmware BCM43430/1 version 7.45.41.46 (r666254 CY) which was supplied for RPI3 also works for RPI02W. And there is a patch in nexmon for this firmware version.
> 
> You can found original 7.45.41.46 firmware version in this package: http://archive.raspberrypi.org/debian/pool/main/f/firmware-nonfree/firmware-brcm80211_0.43+rpi6_all.deb
> 
> This firmware has some bug which sometimes may lead to not responding state. But this is better than nothing.

Any steps or guide on how to execute it?

**Comment by Xendr1k** [comment]
Any updates? I have the same problem with the same version of firmware.
dmesg | grep "Firmware: BCM43430"
[   10.316434] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM43430/1 wl0: Feb  5 2021 12:49:07 version 7.45.96 (r745790) FWI     D 01-71817851 es7.
I would like to be able to make it work in monitor mode

**Comment by rudyrdx** [comment]
> Any updates? I have the same problem with the same version of firmware. dmesg | grep "Firmware: BCM43430" [ 10.316434] brcmfmac: brcmf_c_preinit_dcmds: Firmware: BCM43430/1 wl0: Feb 5 2021 12:49:07 version 7.45.96 (r745790) FWI D 01-71817851 es7. I would like to be able to make it work in monitor mode

like @qrp73 suggested, first download rpi os lite, downgrade the firmware, then setup nexmon.
I have created a list of commands to execute to achieve this in one of my Rpi02w repo you can look it up.

**Comment by SmillerMP** [comment]
Hi, I think I found a solution for Kali Pi Zero 2W

as @qrp73 mentioned, is necesary change the firmware, to do that run:

` wget http://archive.raspberrypi.org/debian/pool/main/f/firmware-nonfree/firmware-brcm80211_0.43+rpi6_all.deb`
`sudo apt install ./firmware-brcm80211_0.43+rpi6_all.deb -y --allow-downgrades`

This will give some warning but it is not an error

To make kali pi run this firmware instead of the original one run this:

Backup the orginal firmware:
`sudo cp /lib/firmware/brcm/brcmfmac43436s-sdio.bin /lib/firmware/brcm/brcmfmac43436s-sdio.bin.bak`
`sudo cp /lib/firmware/brcm/brcmfmac43436s-sdio.bin /lib/firmware/brcm/brcmfmac43436s-sdio.bin.bak`

Make a symlink to the new firmware
`sudo ln -sf /lib/firmware/brcm/brcmfmac43430-sdio.bin /lib/firmware/brcm/brcmfmac43436s-sdio.bin`
`sudo ln -sf /lib/firmware/brcm/brcmfmac43430-sdio.txt /lib/firmware/brcm/brcmfmac43436s-sdio.txt`
`reboot`

and now it should allow using monitor mode, from what I see the bcm43430a1 driver is well configured so it will not be necessary to recompile it

**Comment by SmillerMP** [comment]
do what I mentioned above, install the firmware and compile the driver for bcm43430a1/7_45_41_46, with `iw dev wlan0 info | gawk '/wiphy/ {printf "phy" $2}'` you activate a new interface in monitor mode

**Comment by utkarsharora100** [comment]
![Image](https://github.com/user-attachments/assets/2607a591-577c-4427-8242-cffe582f889e)
@SmillerMP Thanks for the reply!
Is this some sort of bug?
The clm_blob file is only missing for brcmfmac43436s.

**Comment by utkarsharora100** [comment]
![Image](https://github.com/user-attachments/assets/1149828e-948f-47fd-819b-3e6c916de90f)
@SmillerMP Thanks for the reply!
I followed the steps given above and still I am not able to use monitor mode. It says no clm_blob available even though I have set the country code as IN.

**Comment by SmillerMP** [comment]
No problem men
as far i can see, you have another version of firmware, you have the 7.45.96 version

you must have the 7.45.41.46 version, this is very important 

![Image](https://github.com/user-attachments/assets/43b8d7b3-73d7-468c-9db1-48a5a05f0068)

@rudyrdx  have a worksteps to put the interface on monitor mode, check this [](https://github.com/rudyrdx/Raspberry-Pi-Zero-2w/blob/main/Monitor%20Mode.md)


if you have the correct version driver for BCM43430, you only need to install the 7.45.41.46 firmware, the steps can be a little different on kali but you can tell me if you have any problems or if you need help with something.

**Comment by utkarsharora100** [comment]
It worked finally. Thanks a lot 🫡
Earlier I had ignored this error.

Notice: Download is performed unsandboxed as root as file '/home/utkarsh/firmwar e-brcm80211_0.43+rpi6_all.deb' couldn't be accessed by user '_apt'. - pkgAcquire ::Run (13: Permission denied)

To fix this I followed these 2 steps on the terminal -->
1) Write `sudo vi /etc/apt/apt.conf.d/10sandbox`.
2)  Add the following line `APT::Sandbox::User "root";` then write and quit.

<img width="598" alt="Image" src="https://github.com/user-attachments/assets/394278d7-8977-4e2f-859d-9e0c93f3efd2" />

**Comment by utkarsharora100** [comment]
It worked finally. Thanks a lot 🫡
Earlier I had ignored this error.

Notice: Download is performed unsandboxed as root as file '/home/utkarsh/firmwar e-brcm80211_0.43+rpi6_all.deb' couldn't be accessed by user '_apt'. - pkgAcquire ::Run (13: Permission denied)

To fix this I followed these 2 steps on the terminal -->

1) Write sudo vi /etc/apt/apt.conf.d/10sandbox.
2) Add the following line APT::Sandbox::User "root"; then write and quit.

<img width="598" alt="Image" src="https://github.com/user-attachments/assets/c042f9a4-a2fb-4c10-b06e-4b509cea2c34" />

**Comment by SmillerMP** [comment]
Perfect, you're welcome. Have a happy hacking
